### PR TITLE
Refactor & Shadowlands

### PR DESCRIPTION
**Pull request by LiangYuxuan**
#397 

I've done some refactor stuffs but not all, remaining history, tooltip and so on. However, Shadowlands is coming so I managed to get SI work on PTR (no beta :( ), and let rest of refactor stuffs delayed.

Due to breaking API changes, this branch is **NOT** compatible with BfA. So I package this branch with libs but not locales.

[SavedInstances-Shadowlands-e9e35bd.zip](https://github.com/SavedInstances/SavedInstances/files/5156664/SavedInstances-Shadowlands-e9e35bd.zip)

It'll be very helpful if anyone could test this on beta. :)

### CODE REVIEW

**Comment by Stanzilla** [comment]
Oh wow, nice job! This will take a while to review and test!

**Comment by LiangYuxuan** [comment]
Hey @Stanzilla , would you like to grap WeakAura's GitHub actions to build artifact for this? ;)

**Comment by Stanzilla** [comment]
Oh good idea, I can see if I have time for that, been super busy at work atm

**Comment by LiangYuxuan** [comment]
updated 4a41f0f to fix dummy parent id, and this is ready to pre-patch if no bug is reported.

still get some change before launch, but not in this PR I think:
* new world boss quest id, [3 out of 4 quests](https://shadowlands.wowhead.com/quests/type:112?filter=35;9;0) added in database now
* [alchemy transmute](https://www.wowhead.com/guides/shadowlands-alchemy-profession#transmutes-and-alchemy-stone), still in placeholder
* weekend event quest id, still not exists
* mythic plus keystone abbr., wait for RIO ;)
* mythic plus keystone item id ([180653](https://shadowlands.wowhead.com/item=180653/mythic-keystone)), won't update before launch to support BfA's keystone in pre-patch

**Comment by LiangYuxuan** [comment]
updated LFR Castle Nathria Wing 3 name from build 35755

**Comment by Stanzilla** [comment]
Okay which parts are not backwards compatible? Otherwise we could start testing in alpha now and do a release on pre-patch.

**Comment by LiangYuxuan** [comment]
All commits starting with "compatibility", and f211e0b, using SL APIs. Also new SL stuffs are added into currency, LFR, paragon and tradeskill. It's ready to test :) (in beta and ptr of course)

**Comment by Stanzilla** [comment]
Okay so I guess the only nit I have regarding the new module structure that I'd rather keep the upvalues generic, so 

```lua
local SI, L = unpack(select(2, ...))
local Module = SI:NewModule('Currency', 'AceEvent-3.0', 'AceTimer-3.0', 'AceBucket-3.0')
```

File: SavedInstances/Modules/Progress.lua [hunk 1]
```diff
@@ -22,12 +20,12 @@ local READY_CHECK_READY_TEXTURE = READY_CHECK_READY_TEXTURE
 local READY_CHECK_WAITING_TEXTURE = READY_CHECK_WAITING_TEXTURE
 
 local CONQUEST_QUESTLINE_ID = 782
-local maxLvl = MAX_PLAYER_LEVEL_TABLE[#MAX_PLAYER_LEVEL_TABLE]
+local maxLvl = GetMaxLevelForPlayerExpansion()
 
 -- GLOBAL
```
Comment (Stanzilla) [inline SavedInstances/Modules/Progress.lua:25]: leftover?

**Comment by Stanzilla** [comment]
Logged in to a premade and had a Cata random dungeon lockout 
image

**Comment by Stanzilla** [comment]
and I guess we need new defaults

**Comment by LiangYuxuan** [comment]
Got some time to check LFD daily rewards on live and ptr, currently random classic & timewalking (MOP timewalking still on in CN realm) are not marked donetoday. So they're removed from here.
But I don't have lv85 character in live and random normal/heroic cata & Hour of Twilight queues are disabled on ptr, so I'm not sure about at what level should normal cata & HoT should be ignored, at lease they're not at level 50 :)

**Comment by Stanzilla** [comment]
But this is looking really good! I think we are super ready for prepatch, awesome work!

**Comment by Stanzilla** [comment]
There is this table of level conversion if it helps https://shadowlands.wowhead.com/news=317295/level-squish-in-shadowlands-what-level-will-you-be-after-the-squish

**Comment by LiangYuxuan** [comment]
Based on database [LFGDungeons](https://wow.tools/dbc/?dbc=lfgdungeons&build=9.0.2.35854#page=1), random cata heroic & HoT should be available at level 35. But I think it's not possibile to queue in these since no one wants to play these dungeons, so I ignore HoT either. Random normal cata should be available at level 10-50 if player choose cata to level, but currently is disabled on ptr so keep it not ignored.

**Comment by Stanzilla** [comment]
We're currently still using the CurseForge packager, I can switch us over to the BigWigs one in a bit

**Comment by LiangYuxuan** [comment]
That's great! Curse's package is not that reliable and we need BigWigs one to build artifact for PRs.

**Comment by Stanzilla** [comment]
Once again, awesome job!

**Pull request merged**